src/alley/wp/alleyvate/features/class-full-page-cache-404.php

```diff
@@ -0,0 +1,260 @@
+
+		if ( headers_sent() ) {
+			return;
+		}
+
+		if ( ! $stale && 'HIT' === $type ) {
+			header( 'X-Alleyvate-404-Cache: HIT' );
+		} elseif ( $stale && 'HIT' === $type ) {
+			header( 'X-Alleyvate-404-Cache: HIT (stale)' );
+		} elseif ( 'MISS' === $type ) {
+			header( 'X-Alleyvate-404-Cache: MISS' );
+		}
+	}
+
+	/**
+	 * Start output buffering, so we can cache the 404 page.
+	 */
+	public static function action__wp(): void {
+		if ( isset( $_SERVER['REQUEST_URI'] ) && self::TEMPLATE_GENERATOR_URI === $_SERVER['REQUEST_URI'] ) {
+			global $wp_query;
+			if ( ! $wp_query->is_404() ) {
+				return;
+			}
+			ob_start( [ self::class, 'finish_output_buffering' ] );
+		}
+	}
+
+	/**
+	 * Finish output buffering.
+	 *
+	 * @param string $buffer Buffer.
+	 * @return string
+	 */
+	public static function finish_output_buffering( string $buffer ): string {
+		global $wp_query;
+		if ( ! $wp_query->is_404() ) {
+			return $buffer;
+		}
+		if ( is_user_logged_in() ) {
+			return $buffer;
+		}
+		if ( ! self::get_cache() && ! empty( $buffer ) ) {
+			self::set_cache( $buffer );
+		}
+		return $buffer;
+	}
+
+	/**
+	 * Get cache.
+	 *
+	 * @return mixed
+	 */
+	public static function get_cache(): mixed {
+		return wp_cache_get( self::CACHE_KEY, self::CACHE_GROUP );
+	}
+
+	/**
+	 * Get stale cache.
+	 *
+	 * @return mixed
+	 */
+	public static function get_stale_cache(): mixed {
+		return wp_cache_get( self::STALE_CACHE_KEY, self::CACHE_GROUP );
+	}
+
+	/**
+	 * Set cache.
+	 *
+	 * @param string $buffer The Output Buffer.
+	 */
+	public static function set_cache( string $buffer ): void {
+		wp_cache_set( self::CACHE_KEY, $buffer, self::CACHE_GROUP, self::CACHE_TIME ); // phpcs:ignore WordPressVIPMinimum.Performance.LowExpiryCacheTime.CacheTimeUndetermined
+		wp_cache_set( self::STALE_CACHE_KEY, $buffer, self::CACHE_GROUP, self::STALE_CACHE_TIME ); // phpcs:ignore WordPressVIPMinimum.Performance.LowExpiryCacheTime.CacheTimeUndetermined
+	}
+
+	/**
+	 * Delete cache.
+	 */
+	public static function delete_cache(): void {
+		wp_cache_delete( self::CACHE_KEY, self::CACHE_GROUP );
+		wp_cache_delete( self::STALE_CACHE_KEY, self::CACHE_GROUP );
+	}
+
+	/**
+	 * Prepare response.
+	 *
+	 * @param string $content The content.
+	 *
+	 * @return string
+	 */
+	public static function prepare_response( string $content ): string {
+		// To avoid analytics issues, replace the Generator URI with the requested URI.
+		$uri     = sanitize_text_field( $_SERVER['REQUEST_URI'] ?? '' ); // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+		$content = str_replace(
+			[
+				self::TEMPLATE_GENERATOR_URI,
+				wp_json_encode( self::TEMPLATE_GENERATOR_URI ),
+				esc_html( self::TEMPLATE_GENERATOR_URI ),
+				esc_url( self::TEMPLATE_GENERATOR_URI ),
+			],
+			[
+				$uri,
+				wp_json_encode( $uri ),
+				esc_html( $uri ),
+				esc_url( $uri ),
+			],
+			$content
+		);
+		return $content;
```

Comment:
I _think_ this also works.
```suggestion
		return str_replace(
			[
				self::TEMPLATE_GENERATOR_URI,
				wp_json_encode( self::TEMPLATE_GENERATOR_URI ),
				esc_html( self::TEMPLATE_GENERATOR_URI ),
				esc_url( self::TEMPLATE_GENERATOR_URI ),
			],
			[
				$uri,
				wp_json_encode( $uri ),
				esc_html( $uri ),
				esc_url( $uri ),
			],
			$content
		);
```